### PR TITLE
Initial timer implementation

### PR DESCRIPTION
**Pull request by jeremyjh**

### CODE REVIEW

File: examples/echoplus.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                    let ev = mioco.select_read_from(&[conn.index(), five.index()]);
+                    if ev.index() == conn.index() {
+                        let size = try!(conn.read(&mut buf));
+                        if size == 0 {
+                            /* eof */
+                            break;
+                        }
+                        try!(conn.write_all(&mut buf[0..size]));
+                    } else {
+                        println!("Too slow!.");
```
Comment (dpc) [inline examples/echoplus.rs]: `println` is blocking IO so maybe we should avoid it? Can we just close the connection on the timeout + comment? Easy to test, easy to understand.

File: src/lib.rs [hunk 2]
```diff
@@ -204,6 +204,67 @@ where T:Reflect+'static {
...
+        self as &Any
+    }
+
+    fn as_any_mut(&mut self) -> &mut Any {
+        self as &mut Any
+    }
+
+    fn register(&self, event_loop : &mut EventLoop<Server>, token : Token, _interest : EventSet) {
+        let mut state = self.state.borrow_mut();
+        if state.status != TimerStatus::Stopped {
```
Comment (dpc) [inline src/lib.rs]: `assert` or maybe a `debug_assert` ?

File: src/lib.rs [hunk 3]
```diff
@@ -453,15 +518,15 @@ where T : Reflect+'static {
     }
 
     /// Access raw mio type
-    pub fn with_raw<F>(&self, f : F)
-        where F : Fn(&T) {
+    pub fn with_raw<F,R>(&self, f : F) -> R
```
Comment (dpc) [inline src/lib.rs]: Space before `R`.

File: src/lib.rs [hunk 4]
```diff
@@ -1100,6 +1261,78 @@ where T : Reflect+'static {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum TimerStatus  {
+    Running,
+    Stopped,
+    Fired,
```
Comment (dpc) [inline src/lib.rs]: Change `Fired` to ... hmmm ... `Done`/`Complete`/`Ready`?

File: src/lib.rs [hunk 5]
```diff
@@ -1100,6 +1261,78 @@ where T : Reflect+'static {
...
+
+struct TimerState {
+    handle: Option<mio::Timeout>,
+    status: TimerStatus,
+    iteration: u32
+}
+
+/// A Timer is used in `select` functions to create a timeout event.
+pub struct Timer {
+    //Delay in ms
```
Comment (dpc) [inline src/lib.rs]: Space between `//` and `D`. Generally I'm flexible with formatting, but consistency is good.

File: src/lib.rs [hunk 6]
```diff
@@ -1100,6 +1261,78 @@ where T : Reflect+'static {
...
+    Running,
+    Stopped,
+    Fired,
+    Canceled,
+}
+
+struct TimerState {
+    handle: Option<mio::Timeout>,
+    status: TimerStatus,
+    iteration: u32
```
Comment (dpc) [inline src/lib.rs]: TODO for later: Merge this with with `tick` counter somehow.

**Comment by dpc** [comment]
I think it would be better if you just squashed my patches into your change so it's a one self-contained feature-adding fix. I'm still unsure about API details, but it looks good enough to merge and we can bikesched afterwards.

Let me know about the squashing, fixing spacing and minor fixes. I can just fix them along the way later.

I think Timers are making mioco feature complete for 0.1 and I'm very happy about it! Thanks!

**Comment by jeremyjh** [comment]
Ok - I will fix this up - thanks for all the comments. Just to 2x check - I did squash your two final commits but you want me to squash the wakups branch changes too?

**Comment by dpc** [comment]
Ah. No. Sorry, my mistake.

**Comment by jeremyjh** [comment]
Ok fixed up per your comments and re-squashed.

**Pull request merged**